lua/cfc_disconnect_interface/client/cl_detached_timer.lua

```diff
@@ -0,0 +1,52 @@
+dTimer = {}
+
+dTimer.timers = {}
+dTimer.idCounter = 0
+
+function dTimer.Create( id, delay, reps, f )
+    dTimer.timers[id] = {id = id, delay = delay, reps = reps, f = f, lastCall = SysTime()}
+end
+
+function dTimer.Adjust( id, delay, reps, f )
+    if dTimer.timers[id] then
+        dTimer.timers[id].delay = delay or dTimer.timers[id].delay
+        dTimer.timers[id].reps = reps or dTimer.timers[id].reps
+        dTimer.timers[id].f = f or dTimer.timers[id].f
+        return true
+    end
+    return false
+end
+
+function dTimer.Exists( id )
+    return not not dTimer.timers[id]
+end
+
+function dTimer.Remove( id )
+    dTimer.timers[id] = nil
+end
+
+function dTimer.RepsLeft( id )
+    return dTimer.timers[id] and dTimer.timers[id].reps or -1
+end
+
+function dTimer.Simple( delay, f )
+    dTimer.Create( "SimpleTimer" .. dTimer.idCounter, delay, 1, f )
+    dTimer.idCounter = dTimer.idCounter + 1
+end
+
+
+hook.Add( "Think", "cfc_di_detatched_timer", function()
+    local s = SysTime()
+    for k, t in pairs( dTimer.timers ) do
+        if s - t.lastCall > t.delay then
+            t.lastCall = s
+            if t.reps > 0 then
+                t.reps = t.reps - 1
+                if t.reps == 0 then
+                    dTimer.timers[k] = nil
+                end
+            end
+            t.f()
+        end
+    end
```

Comment:
This section would greatly benefit from some helpful variable names and some better structuring.

i.e.
```lua
local poppedDelay = s - t.lastCall > t.delay

continue if not poppedDelay

t.lastCall = s
local hasReps = t.reps > 0

continue if not hasReps

t.reps = t.reps - 1

hasReps = t.reps != 0

continue if hasReps

local function clearTimer( index )
    dTimer.timers[index] = nil
end

clearTimer(k)
```

Reply:
Not sure it's necessary to extract single use conditions into variables, thoughts on this:
```
hook.Add( "Think", "cfc_di_detatched_timer", function()
    local s = SysTime()
    for k, t in pairs( dTimer.timers ) do
        if s - t.lastCall <= t.delay then continue end -- Not yet past delay
        t.lastCall = s
        t.func()
        if not t.reps > 0 then continue end -- No need to decrement reps
        t.reps = t.reps - 1
        if t.reps == 0 then
            dTimer.Remove( k )
        end
    end
end )
```

Reply:
replace `not t.reps > 0` with `t.reps == 0`, oops

Reply:
Not a huge fan of extracting single use conditions into vars, thoughts on this:
```
hook.Add( "Think", "cfc_di_detatched_timer", function()
    local time = SysTime()
    for id, curTimer in pairs( dTimer.timers ) do
        if time - curTimer.lastCall <= curTimer.delay then continue end -- Not yet past delay
        curTimer.lastCall = time
        curTimer.func()
        if curTimer.reps == 0 then continue end -- No need to decrement reps
        curTimer.reps = curTimer.reps - 1
        if curTimer.reps == 0 then
            dTimer.Remove( id )
        end
    end
end )
```

Reply:
Single use vars tell the reader what you're thinking as you go through it without needing to add comments, it's a self-documentation strategy

Reply:
But then ye get ugly code :( Do you want me to change this then? or are the comments ok

Reply:
I don't think it's ugly at all, and there are lots of ways to make it look nice compared to the hasty way I wrote it. I definitely don't like the comments on the same line, though